pkg/ring/client/pool.go

```diff
@@ -0,0 +1,198 @@
+package grpcclient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+	"google.golang.org/grpc/health/grpc_health_v1"
+
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// PoolClient is the interface that should be implemented by a
+// client managed by the pool.
+type PoolClient interface {
+	grpc_health_v1.HealthClient
+	io.Closer
+}
+
+// PoolFactory defines the signature for a client factory.
+type PoolFactory func(addr string) (PoolClient, error)
+
+// PoolConfig is config for creating a Pool.
+type PoolConfig struct {
+	CheckInterval      time.Duration
+	HealthCheckEnabled bool
+	HealthCheckTimeout time.Duration
+}
+
+// Pool holds a cache of grpc_health_v1 clients.
+type Pool struct {
+	services.Service
+
+	cfg         PoolConfig
+	ring        ring.ReadRing
+	factory     PoolFactory
+	logger      log.Logger
+	serviceName string
```

Comment:
I would call this `clientName` or `clientType`, to avoid confusion with this being a pool's name (since it's a service).